File: python/src/prts_mcp/data/story_character.py
"""Story character tracking — who appears where, and who speaks in an event.

Split from story.py. Reuses the lazily-cached search index built by
``story_search`` (which already walks every chapter and line), so this module
performs no file IO of its own and needs no separate cache. Two questions are
answered:

- *Where does a character show up?* — chapters/events where the character
  **speaks** (dialog ``role`` exact match, consistent with the ``character``
  filter of ``search_stories``) or is **mentioned** (name appears as a
  substring in any line's text).
- *Who speaks in an event?* — distinct speakers with their dialog line counts.
"""
from __future__ import annotations

from pathlib import Path

from prts_mcp.data.stores import JsonStore
from prts_mcp.data.story_reader import (
    CharacterAppearance,
    CharacterAppearanceResult,
    SpeakerCount,
    story_store,
)
from prts_mcp.data.story_search import story_search_index, StorySearchChapter


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _classify_chapter(chapter: StorySearchChapter, name_lower: str) -> tuple[bool, bool]:
    """Return (speaks, mentioned) flags for one indexed chapter.

    Args:
        chapter: a ``StorySearchChapter`` from the search index.
        name_lower: the lowercased character name to look for.

    ``speaks`` is True when any dialog line's ``role`` equals *name_lower*
    (case-insensitive), mirroring the ``character`` filter of
    :func:`search_stories`. ``mentioned`` is True when *name_lower* occurs as a
    substring in any line's text.
    """
    speaks = False
    mentioned = False
    for line in chapter.lines:
        if line.type == "dialog" and (line.role or "").lower() == name_lower:
            speaks = True
        if name_lower in line.text.lower():
            mentioned = True
        if speaks and mentioned:
            break
    return speaks, mentioned


# ---------------------------------------------------------------------------
# Public API — zip-path wrappers
# ---------------------------------------------------------------------------


def find_character_appearances(
    zip_path: Path,
    name: str,
    scope: str | None = None,
    max_events: int = 50,
) -> CharacterAppearanceResult:
    """Find chapters/events where *name* speaks or is mentioned.

    Convenience wrapper around :func:`find_character_appearances_from_store`
    that auto-creates a ZipStore from *zip_path*.
    """
    with story_store(zip_path) as store:
        return find_character_appearances_from_store(
            store, name, scope=scope, max_events=max_events,
        )


def build_character_appearances(
    zip_path: Path,
    name: str,
    scope: str | None = None,
    max_events: int = 50,
) -> dict:
    """Build the structured payload for character appearance lookup."""
    with story_store(zip_path) as store:
        return build_character_appearances_from_store(
            store, name, scope=scope, max_events=max_events,
        )


def render_character_appearances(data: dict) -> str:
    """Render a character appearance payload to markdown."""
    appearances = data["appearances"]
    if not appearances:
        scope = data.get("filters", {}).get("scope")
        scope_note = f"（限定活动：{scope!r}）" if scope else ""
        return f"未找到「{data['name']}」的出场记录。{scope_note}"

    parts = [f"# 「{data['name']}」的出场（共 {data['total']} 章）"]
    for ap in appearances:
        tags = []
        if ap["speaks"]:
            tags.append("speaks")
        if ap["mentioned"]:
            tags.append("mentioned")
        tag = "+".join(tags)
        name_disp = f"{ap['story_code']} {ap['story_name']}".strip()
        parts.append(
            f"- [{tag}] {ap['event_id']} / {name_disp}（key: {ap['story_key']}）"
        )
    return "\n".join(parts)


def find_speakers_in(zip_path: Path, event_id: str) -> list[SpeakerCount]:
    """List every speaker in an event, with dialog line counts.

    Convenience wrapper around :func:`find_speakers_in_from_store`.
    """
    with story_store(zip_path) as store:
        return find_speakers_in_from_store(store, event_id)


def build_speakers_in_event(zip_path: Path, event_id: str) -> dict:
    """Build the structured payload for speakers in one event."""
    with story_store(zip_path) as store:
        return build_speakers_in_event_from_store(store, event_id)


def render_speakers_in_event(data: dict) -> str:
    """Render an event speaker-count payload to markdown."""
    speakers = data["speakers"]
    if not speakers:
        return f"活动 {data['event_id']!r} 暂无对话发言者数据。"

    parts = [f"# {data['event_id']} 的发言角色（共 {data['total']} 位）"]
    for sp in speakers:
        parts.append(f"- {sp['name']}（{sp['line_count']} 句）")
    return "\n".join(parts)


# ---------------------------------------------------------------------------
# Public API — store-based variants
# ---------------------------------------------------------------------------


def find_character_appearances_from_store(
    store: JsonStore,
    name: str,
    scope: str | None = None,
    max_events: int = 50,
) -> CharacterAppearanceResult:
    """Find chapters/events where *name* speaks or is mentioned.

    Args:
        store: JsonStore (ZipStore or DirectoryStore) for story data.
        name: Character name (e.g. 「阿米娅」). ``speaks`` uses case-insensitive
            exact match on dialog ``role``; ``mentioned`` uses substring match
            on any line's text. Use the full name to avoid substring false
            positives (e.g. 「阿」 matches 「阿米娅」).
        scope: Optional event_id filter (e.g. "act31side").
        max_events: Cap on returned chapters (default 50, max 200).

    Returns:
        Aggregated appearance report ordered by the index's chapter order.
    """
    if not name:
        raise ValueError("name 不能为空。")
    if max_events < 1:
        raise ValueError("max_events 必须 >= 1。")
    if max_events > 200:
        raise ValueError("max_events 必须 <= 200。")

    index = story_search_index(store)

    if scope is not None and scope not in index.event_ids:
        raise KeyError(f'未找到匹配的活动："{scope}"。')

    name_lower = name.lower()
    appearances: list[CharacterAppearance] = []

    for chapter in index.chapters:
        if scope is not None and chapter.event_id != scope:
            continue
        speaks, mentioned = _classify_chapter(chapter, name_lower)
        if not speaks and not mentioned:
            continue
        appearances.append(CharacterAppearance(
            event_id=chapter.event_id,
            story_key=chapter.story_key,
            story_code=chapter.story_code,
            story_name=chapter.story_name,
            speaks=speaks,
            mentioned=mentioned,
        ))
        if len(appearances) >= max_events:
            break

    return CharacterAppearanceResult(
        name=name,
        total_chapters=len(appearances),
        appearances=appearances,
    )


def build_character_appearances_from_store(
    store: JsonStore,
    name: str,
    scope: str | None = None,
    max_events: int = 50,
) -> dict:
    """Build a character appearance payload using a JSON store."""
    result = find_character_appearances_from_store(
        store, name, scope=scope, max_events=max_events,
    )
    return {
        "name": result.name,
        "total": result.total_chapters,
        "filters": {"scope": scope},
        "appearances": [
            {
                "event_id": ap.event_id,
                "story_code": ap.story_code,
                "story_name": ap.story_name,
                "story_key": ap.story_key,
                "speaks": ap.speaks,
                "mentioned": ap.mentioned,
            }
            for ap in result.appearances
        ],
    }


def find_speakers_in_from_store(
    store: JsonStore,
    event_id: str,
) -> list[SpeakerCount]:
    """List every speaker in an event, with dialog line counts.

    Iterates the cached search index, collecting distinct dialog ``role``
    values (dropping ``None``/empty — those are narration placeholders) and
    counting their lines. Results are sorted by line count descending, then by
    name for a stable order.

    Args:
        store: JsonStore (ZipStore or DirectoryStore) for story data.
        event_id: Event to restrict the scan to (e.g. "act31side").

    Returns:
        Speakers with their dialog line counts, most prolific first.
    """
    index = story_search_index(store)

    if event_id not in index.event_ids:
        raise KeyError(f'未找到匹配的活动："{event_id}"。')

    counts: dict[str, int] = {}
    for chapter in index.chapters:
        if chapter.event_id != event_id:
            continue
        for line in chapter.lines:
            if line.type == "dialog" and line.role:
                counts[line.role] = counts.get(line.role, 0) + 1

    speakers = [
        SpeakerCount(name=name, line_count=count)
        for name, count in counts.items()
    ]
    speakers.sort(key=lambda s: (-s.line_count, s.name))
    return speakers


def build_speakers_in_event_from_store(store: JsonStore, event_id: str) -> dict:
    """Build an event speaker-count payload using a JSON store."""
    speakers = find_speakers_in_from_store(store, event_id)
    return {
        "event_id": event_id,
        "total": len(speakers),
        "speakers": [
            {
                "name": sp.name,
                "line_count": sp.line_count,
            }
            for sp in speakers
        ],
    }

File: ts/src/data/storyCharacter.ts
/**
 * Story character tracking — who appears where, and who speaks in an event.
 *
 * Split from story.ts. Reuses the lazily-cached search index built by
 * storySearch (which already walks every chapter and line), so this module
 * performs no file IO of its own and needs no separate cache. Two questions
 * are answered:
 *
 * - *Where does a character show up?* — chapters/events where the character
 *   **speaks** (dialog `role` exact match, consistent with the `character`
 *   filter of searchStories) or is **mentioned** (name appears as a substring
 *   in any line's text).
 * - *Who speaks in an event?* — distinct speakers with their dialog line
 *   counts.
 *
 * Mirrors python/src/prts_mcp/data/story_character.py.
 */

import { type JsonStore } from "./stores.js";
import {
  type CharacterAppearance,
  type CharacterAppearanceResult,
  type SpeakerCount,
  withStoryStore,
} from "./storyReader.js";
import { storySearchIndex, type StorySearchChapter } from "./storySearch.js";

export interface CharacterAppearancesPayload {
  name: string;
  total: number;
  filters: {
    scope: string | null;
  };
  appearances: Array<{
    event_id: string;
    story_code: string;
    story_name: string;
    story_key: string;
    speaks: boolean;
    mentioned: boolean;
  }>;
}

export interface SpeakersInEventPayload {
  event_id: string;
  total: number;
  speakers: Array<{
    name: string;
    line_count: number;
  }>;
}

function pyRepr(value: string | null | undefined): string {
  if (value === null || value === undefined) return "None";
  return `'${value}'`;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Return [speaks, mentioned] flags for one indexed chapter.
 *
 * `speaks` is true when any dialog line's `role` equals nameLower
 * (case-insensitive), mirroring the `character` filter of searchStories.
 * `mentioned` is true when nameLower occurs as a substring in any line's text.
 */
function classifyChapter(
  chapter: StorySearchChapter,
  nameLower: string,
): [boolean, boolean] {
  let speaks = false;
  let mentioned = false;
  for (const line of chapter.lines) {
    if (line.type === "dialog" && (line.role ?? "").toLowerCase() === nameLower) {
      speaks = true;
    }
    if (line.text.toLowerCase().includes(nameLower)) {
      mentioned = true;
    }
    if (speaks && mentioned) {
      break;
    }
  }
  return [speaks, mentioned];
}

// ---------------------------------------------------------------------------
// Public API — zip-path wrappers
// ---------------------------------------------------------------------------

export function findCharacterAppearances(
  zipPath: string,
  name: string,
  scope?: string,
  maxEvents = 50,
): CharacterAppearanceResult {
  return withStoryStore(zipPath, (store) =>
    findCharacterAppearancesFromStore(store, name, scope, maxEvents),
  );
}

export function buildCharacterAppearances(
  zipPath: string,
  name: string,
  scope?: string,
  maxEvents = 50,
): CharacterAppearancesPayload {
  return withStoryStore(zipPath, (store) =>
    buildCharacterAppearancesFromStore(store, name, scope, maxEvents),
  );
}

export function renderCharacterAppearances(data: CharacterAppearancesPayload): string {
  if (data.appearances.length === 0) {
    const scope = data.filters.scope;
    const scopeNote = scope ? `（限定活动：${pyRepr(scope)}）` : "";
    return `未找到「${data.name}」的出场记录。${scopeNote}`;
  }

  const parts = [`# 「${data.name}」的出场（共 ${data.total} 章）`];
  for (const ap of data.appearances) {
    const tags: string[] = [];
    if (ap.speaks) tags.push("speaks");
    if (ap.mentioned) tags.push("mentioned");
    const nameDisp = `${ap.story_code} ${ap.story_name}`.trim();
    parts.push(`- [${tags.join("+")}] ${ap.event_id} / ${nameDisp}（key: ${ap.story_key}）`);
  }
  return parts.join("\n");
}

export function findSpeakersIn(
  zipPath: string,
  eventId: string,
): SpeakerCount[] {
  return withStoryStore(zipPath, (store) =>
    findSpeakersInFromStore(store, eventId),
  );
}

export function buildSpeakersInEvent(
  zipPath: string,
  eventId: string,
): SpeakersInEventPayload {
  return withStoryStore(zipPath, (store) => buildSpeakersInEventFromStore(store, eventId));
}

export function renderSpeakersInEvent(data: SpeakersInEventPayload): string {
  if (data.speakers.length === 0) {
    return `活动 ${pyRepr(data.event_id)} 暂无对话发言者数据。`;
  }

  const parts = [`# ${data.event_id} 的发言角色（共 ${data.total} 位）`];
  for (const sp of data.speakers) {
    parts.push(`- ${sp.name}（${sp.line_count} 句）`);
  }
  return parts.join("\n");
}

// ---------------------------------------------------------------------------
// Public API — store-based variants
// ---------------------------------------------------------------------------

export function findCharacterAppearancesFromStore(
  store: JsonStore,
  name: string,
  scope?: string,
  maxEvents = 50,
): CharacterAppearanceResult {
  if (!name) throw new Error("name 不能为空。");
  if (maxEvents < 1) throw new Error("max_events 必须 >= 1。");
  if (maxEvents > 200) throw new Error("max_events 必须 <= 200。");

  const index = storySearchIndex(store);

  if (scope !== undefined && !index.eventIds.has(scope)) {
    throw new Error(`未找到匹配的活动：${JSON.stringify(scope)}。`);
  }

  const nameLower = name.toLowerCase();
  const appearances: CharacterAppearance[] = [];

  for (const chapter of index.chapters) {
    if (scope !== undefined && chapter.eventId !== scope) continue;
    const [speaks, mentioned] = classifyChapter(chapter, nameLower);
    if (!speaks && !mentioned) continue;
    appearances.push({
      eventId: chapter.eventId,
      storyKey: chapter.storyKey,
      storyCode: chapter.storyCode,
      storyName: chapter.storyName,
      speaks,
      mentioned,
    });
    if (appearances.length >= maxEvents) break;
  }

  return {
    name,
    totalChapters: appearances.length,
    appearances,
  };
}

export function buildCharacterAppearancesFromStore(
  store: JsonStore,
  name: string,
  scope?: string,
  maxEvents = 50,
): CharacterAppearancesPayload {
  const result = findCharacterAppearancesFromStore(store, name, scope, maxEvents);
  return {
    name: result.name,
    total: result.totalChapters,
    filters: { scope: scope ?? null },
    appearances: result.appearances.map((ap) => ({
      event_id: ap.eventId,
      story_code: ap.storyCode,
      story_name: ap.storyName,
      story_key: ap.storyKey,
      speaks: ap.speaks,
      mentioned: ap.mentioned,
    })),
  };
}

export function findSpeakersInFromStore(
  store: JsonStore,
  eventId: string,
): SpeakerCount[] {
  const index = storySearchIndex(store);

  if (!index.eventIds.has(eventId)) {
    throw new Error(`未找到匹配的活动：${JSON.stringify(eventId)}。`);
  }

  const counts = new Map<string, number>();
  for (const chapter of index.chapters) {
    if (chapter.eventId !== eventId) continue;
    for (const line of chapter.lines) {
      if (line.type === "dialog" && line.role) {
        counts.set(line.role, (counts.get(line.role) ?? 0) + 1);
      }
    }
  }

  const speakers: SpeakerCount[] = [];
  for (const [speakerName, lineCount] of counts) {
    speakers.push({ name: speakerName, lineCount });
  }
  // Sort by line count desc, then by name using codepoint ordering to match
  // the Python implementation's `(-line_count, name)` tuple sort.
  speakers.sort((a, b) => b.lineCount - a.lineCount || (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  return speakers;
}

export function buildSpeakersInEventFromStore(
  store: JsonStore,
  eventId: string,
): SpeakersInEventPayload {
  const speakers = findSpeakersInFromStore(store, eventId);
  return {
    event_id: eventId,
    total: speakers.length,
    speakers: speakers.map((sp) => ({
      name: sp.name,
      line_count: sp.lineCount,
    })),
  };
}
